File: myflow/gofast/sensor/generate.go
package sensor

import (
	_ "embed"
	"encoding/json"
	"fmt"
	"net/url"
	"regexp"
	"strconv"
	"strings"
)

// SensorInput is the public contract (matches script/input.json and babelflow's
// SensorInput). Only Bmsz is required (its first numeric token is the cipher seed);
// the rest fall back to template values.
type SensorInput struct {
	ScriptUrl      string `json:"ScriptUrl"`
	PageUrl        string `json:"PageUrl"`
	UserAgent      string `json:"UserAgent"`
	Abck           string `json:"Abck"`
	Bmsz           string `json:"Bmsz"`
	Version        string `json:"Version"`
	AcceptLanguage string `json:"AcceptLanguage"`
}

// template mirrors template.json. The device profile is fixed; input-derived and
// dynamic fields are substituted at generate time.
type template struct {
	PrefixFixed string `json:"prefixFixed"`
	Section0    string `json:"section0"`
	Meta        string `json:"meta"`
	SeedExample int    `json:"seedExample"`
	OrigInputs  struct {
		UserAgent   string `json:"UserAgent"`
		PageUrlHost string `json:"PageUrlHost"`
		ScriptToken string `json:"ScriptToken"`
		Locale      string `json:"Locale"`
		Bmsz        string `json:"Bmsz"`
	} `json:"origInputs"`
	Plain string `json:"plain"`
}

//go:embed template.json
var templateJSON []byte

var tmpl = func() template {
	var t template
	if err := json.Unmarshal(templateJSON, &t); err != nil {
		panic("sensor: bad embedded template.json: " + err.Error())
	}
	return t
}()

// prefix field layout: <ver>;<a>;<b>;<c>;<seed>;[,…]; — field[4] is the cipher seed.
var prefixRe = regexp.MustCompile(`\d+(;\d+)?;(\d+);(\d+);[\d,]+;`)

const seedField = 4

var (
	scRe     = regexp.MustCompile(`sc:(\d{13})`)
	localeRe = regexp.MustCompile(`^\s*([A-Za-z]{2}(?:-[A-Za-z]{2,})?)`)
	digitsRe = regexp.MustCompile(`^\d+$`)
	segOK    = regexp.MustCompile(`^[A-Za-z0-9_-]+$`)
)

func atoiOr(s string, def int) int {
	if n, err := strconv.Atoi(s); err == nil {
		return n
	}
	return def
}

// bmszSeed extracts the cipher seed: the first numeric '~'-token, e.g.
// "<hex>~3158840~4605490" -> 3158840. (Not a regex over the whole string — the
// leading hex contains digit runs.)
func bmszSeed(bmsz string) (int, error) {
	parts := strings.Split(bmsz, "~")
	for _, p := range parts[1:] {
		if digitsRe.MatchString(p) {
			return strconv.Atoi(p)
		}
	}
	if len(parts) > 0 && digitsRe.MatchString(parts[0]) {
		return strconv.Atoi(parts[0])
	}
	return 0, fmt.Errorf("cannot extract bm_sz seed token from: %s", bmsz)
}

// scriptToken picks the script-path token from ScriptUrl, matching the template
// token's shape (>=12 chars, [A-Za-z0-9_-]); falls back to the template token.
func scriptToken(scriptURL, tmplToken string) string {
	u, err := url.Parse(scriptURL)
	if err != nil {
		return tmplToken
	}
	for _, s := range strings.Split(u.Path, "/") {
		if len(s) >= 12 && segOK.MatchString(s) {
			return s
		}
	}
	return tmplToken
}

// pageHost strips the scheme (the template stores host+path without scheme).
func pageHost(pageURL string) string {
	return regexp.MustCompile(`^https?://`).ReplaceAllString(pageURL, "")
}

// primaryLocale extracts the primary locale: "en-US,en;q=0.9" -> "en-US".
func primaryLocale(acceptLanguage string) string {
	m := localeRe.FindStringSubmatch(acceptLanguage)
	if len(m) > 1 {
		return m[1]
	}
	return ""
}

// Generate builds a v3 sensor_data body from input, using nowMs as the refreshed
// `sc` ms-epoch timestamp. Mirrors generate.js generate().
func Generate(input SensorInput, nowMs int64) (string, error) {
	plain := tmpl.Plain
	o := tmpl.OrigInputs

	// 1) input-derived substitutions (verbatim string replace)
	if input.UserAgent != "" {
		plain = strings.ReplaceAll(plain, o.UserAgent, input.UserAgent)
	}
	if input.PageUrl != "" {
		plain = strings.ReplaceAll(plain, o.PageUrlHost, pageHost(input.PageUrl))
	}
	if input.ScriptUrl != "" {
		plain = strings.ReplaceAll(plain, o.ScriptToken, scriptToken(input.ScriptUrl, o.ScriptToken))
	}
	if input.AcceptLanguage != "" && o.Locale != "" {
		if loc := primaryLocale(input.AcceptLanguage); loc != "" {
			plain = strings.ReplaceAll(plain, o.Locale, loc)
		}
	}

	// 2) refresh the confirmed dynamic timestamp (`sc`). Other time/counter fields
	// (dau, tid, …) have unknown derivations and stay at template values for now.
	plain = scRe.ReplaceAllString(plain, "sc:"+strconv.FormatInt(nowMs, 10))

	// 3) seed from Bmsz, assemble + encrypt
	seed, err := bmszSeed(input.Bmsz)
	if err != nil {
		return "", err
	}
	body := encodeV3(tmpl.Section0, tmpl.Meta, seed, plain)
	return tmpl.PrefixFixed + strconv.Itoa(seed) + ";" + body, nil
}

// epochRe matches a maximal digit run; Regen filters for 13-digit ms-epochs by value.
var digitRunRe = regexp.MustCompile(`\d+`)

// refreshEpochs replaces every 13-digit ms-epoch-looking number (1[78]xxxxxxxxxxx,
// i.e. 2023–2027) with nowMs. We target by VALUE, not by the `sc:` label, because the
// transposed plaintext de-correlates labels from values (the `sc:` label sits next to
// junk while the real timestamp is in another slot). delta keeps relative offsets if
// you later want multiple timestamps staggered; for now all move to nowMs.
func refreshEpochs(plain string, nowMs int64) (string, int) {
	now := strconv.FormatInt(nowMs, 10)
	n := 0
	out := digitRunRe.ReplaceAllStringFunc(plain, func(run string) string {
		if len(run) == 13 && (run[0] == '1') && (run[1] == '7' || run[1] == '8') {
			n++
			return now
		}
		return run
	})
	return out, n
}

var anchorEpochRe = regexp.MustCompile(`1[78]\d{11}`)

// ShiftAnchor implements the coherent-replay transform (TIMING.md): keep section0/seed
// and all relative timers/behavioral offsets, and shift ONLY the session-anchor epoch by
// deltaMs. Replaying a captured body at time T with deltaMs = T − originalSend keeps
// (T − anchorNew) == originalElapsed, so the elapsed field and offsets stay coherent —
// one field changed instead of the whole timing group. Returns body, old/new anchor, and
// how many anchor occurrences were shifted (expect 1). Risk to verify live: if section0
// binds the anchor, shifting it invalidates section0 (then fall back to timer-shifting).
func ShiftAnchor(sensorData string, deltaMs int64) (string, int64, int64, int, error) {
	r := DecodeV3(sensorData)
	if !r.OK {
		return "", 0, 0, 0, fmt.Errorf("decode capture: %s", r.Reason)
	}
	anchorStr := anchorEpochRe.FindString(r.Plain)
	if anchorStr == "" {
		return "", 0, 0, 0, fmt.Errorf("no anchor epoch (1[78]xxxxxxxxxxx) found in plaintext")
	}
	anchor, _ := strconv.ParseInt(anchorStr, 10, 64)
	newAnchor := anchor + deltaMs
	newStr := strconv.FormatInt(newAnchor, 10)
	plain := strings.ReplaceAll(r.Plain, anchorStr, newStr)
	n := strings.Count(r.Plain, anchorStr)
	return r.Prefix + encodeV3(r.Section0, r.Meta, r.Seed, plain), anchor, newAnchor, n, nil
}

// Regen is capture-assisted generation: decode a real captured sensor_data, keep its
// section0/meta/seed (section0 is server-checked and session-stable, so it must be a
// real one), refresh the dynamic ms-epoch timestamp(s) to nowMs, and re-encode. Unlike
// Generate (template/from-scratch, whose stale section0 fails the check), this reuses a
// valid captured fingerprint. Returns the body and how many timestamps were refreshed.
func Regen(sensorData string, nowMs int64) (string, int, error) {
	r := DecodeV3(sensorData)
	if !r.OK {
		return "", 0, fmt.Errorf("decode capture: %s", r.Reason)
	}
	plain, n := refreshEpochs(r.Plain, nowMs)
	return r.Prefix + encodeV3(r.Section0, r.Meta, r.Seed, plain), n, nil
}

File: myflow/babelflow/generator/generate.js
/*
 * generator/generate.js — Generate(SensorInput) -> v3 sensor_data, from scratch.
 *
 * Architecture (see FINDINGS.md): the 8 SensorInput fields do NOT contain the
 * device fingerprint, so we use a fixed DEVICE PROFILE captured once (template.json,
 * decoded from a real first-POST sensor) and substitute in the input-derived +
 * dynamic fields, then encrypt with the recovered cipher (cipher.js).
 *
 *   sensor_data = "3;0;1;0;<bmszSeed>;" + encodeV3(section0, meta, encryptSub(plain, seed))
 *
 * Input mapping (recovered): Bmsz->cipher seed, UserAgent->adp/version,
 * PageUrl->url field, ScriptUrl->path token, AcceptLanguage->lang; timestamps
 * are refreshed each call. Abck/IP are not part of the sensor body.
 *
 * Usage:
 *   node generate.js '{"ScriptUrl":"...","PageUrl":"...","UserAgent":"...","Bmsz":"36F5..~3158840~..","AcceptLanguage":"en-US","Abck":"...","IP":"..."}'
 *   node generate.js --selftest      # regenerate the template's own inputs == original
 */
const fs = require('fs');
const path = require('path');
const C = require('./cipher');

const TEMPLATE = JSON.parse(fs.readFileSync(path.join(__dirname, 'template.json'), 'utf8'));

// bm_sz seed = the first numeric `~`-token, e.g. "<hex>~3158840~4605490" -> 3158840.
// (NOT a regex over the whole string: the leading hex contains digit runs.)
function bmszSeed(bmsz) {
    const parts = String(bmsz).split('~');
    const tok = parts.slice(1).find(p => /^\d+$/.test(p)) || (/^\d+$/.test(parts[0]) ? parts[0] : null);
    if (!tok) throw new Error('cannot extract bm_sz seed token from: ' + bmsz);
    return parseInt(tok, 10);
}

// the script-path token used in the sensor (a path segment of ScriptUrl). We map
// by matching the template token's position among ScriptUrl path segments.
function scriptToken(scriptUrl, tmplToken) {
    try {
        const segs = new URL(scriptUrl).pathname.split('/').filter(Boolean);
        // template token length ~18 alnum; pick the segment most similar in shape
        const cand = segs.find(s => s.length >= 12 && /^[A-Za-z0-9_-]+$/.test(s));
        return cand || tmplToken;
    } catch (_) { return tmplToken; }
}

function pageHost(pageUrl) {
    // template stores host+path without scheme (matches how it appears in plain)
    return String(pageUrl).replace(/^https?:\/\//, '');
}

// primary locale from an Accept-Language header: "en-US,en;q=0.9" -> "en-US"
function primaryLocale(acceptLanguage) {
    const m = /^\s*([A-Za-z]{2}(?:-[A-Za-z]{2,})?)/.exec(String(acceptLanguage));
    return m ? m[1] : null;
}

function generate(input, opts = {}) {
    const now = opts.now || Date.now();
    let plain = TEMPLATE.plain;
    const o = TEMPLATE.origInputs;

    // 1) input-derived substitutions (verbatim string replace)
    if (input.UserAgent) plain = plain.split(o.UserAgent).join(input.UserAgent);
    if (input.PageUrl) plain = plain.split(o.PageUrlHost).join(pageHost(input.PageUrl));
    if (input.ScriptUrl) plain = plain.split(o.ScriptToken).join(scriptToken(input.ScriptUrl, o.ScriptToken));
    if (input.AcceptLanguage && o.Locale) {
        const loc = primaryLocale(input.AcceptLanguage);
        if (loc) plain = plain.split(o.Locale).join(loc);
    }

    // 2) refresh dynamic timestamps. Only `sc` is a confirmed ms-epoch field;
    // other time/counter fields (dau, tid, ...) have unknown derivations, so we
    // leave them at the template values for now (refine as they're decoded).
    plain = plain.replace(/sc:(\d{13})/g, 'sc:' + now);

    // 3) seed from Bmsz, assemble + encrypt
    const seed = bmszSeed(input.Bmsz);
    const body = C.encodeV3({ section0: TEMPLATE.section0, meta: TEMPLATE.meta, seed, plain });
    return TEMPLATE.prefixFixed + seed + ';' + body;
}

function selftest() {
    // regenerate with the template's own inputs + frozen timestamp -> must equal
    // the exact body we captured (proves substitution + cipher are faithful).
    const o = TEMPLATE.origInputs;
    const frozen = parseInt(/sc:(\d{13})/.exec(TEMPLATE.plain)[1], 10);
    const sd = generate({
        UserAgent: o.UserAgent, PageUrl: 'https://' + o.PageUrlHost,
        ScriptUrl: 'https://www.zalando.pl/x/y/' + o.ScriptToken + '/z',
        Bmsz: 'X~' + TEMPLATE.seedExample + '~Y',
    }, { now: frozen });
    const expectBody = C.encodeV3({ section0: TEMPLATE.section0, meta: TEMPLATE.meta, seed: TEMPLATE.seedExample, plain: TEMPLATE.plain });
    const expect = TEMPLATE.prefixFixed + TEMPLATE.seedExample + ';' + expectBody;
    const ok = sd === expect;
    console.log('selftest (regenerate template inputs == original):', ok ? 'PASS' : 'FAIL');
    if (!ok) {
        for (let i = 0; i < Math.max(sd.length, expect.length); i++) {
            if (sd[i] !== expect[i]) { console.log('  first diff @', i, JSON.stringify(sd.slice(i - 10, i + 10)), 'vs', JSON.stringify(expect.slice(i - 10, i + 10))); break; }
        }
    }
    // round-trip decode sanity
    const dec = C.decodeV3(sd);
    console.log('round-trip decode ok:', dec.ok && dec.plain.includes('"adp"'));
    process.exit(ok ? 0 : 1);
}

// Resolve the SensorInput JSON from: --file <path> | a path arg | '-' (stdin) |
// an inline JSON string. File/stdin avoid PowerShell quoting headaches.
function readInputArg() {
    const args = process.argv.slice(2).filter(a => a !== '--selftest');
    const fi = args.indexOf('--file');
    if (fi !== -1 && args[fi + 1]) return fs.readFileSync(args[fi + 1], 'utf8');
    const a = args[0];
    if (!a) return null;
    if (a === '-') return fs.readFileSync(0, 'utf8');           // stdin
    if (!a.trim().startsWith('{') && fs.existsSync(a)) return fs.readFileSync(a, 'utf8');
    return a;                                                   // inline JSON
}

if (require.main === module) {
    if (process.argv.includes('--selftest')) { selftest(); }
    else {
        const raw = readInputArg();
        if (!raw) {
            console.error([
                'usage:',
                '  node generate.js --file input.json     # easiest on PowerShell',
                '  Get-Content input.json | node generate.js -',
                "  node generate.js '<inline SensorInput JSON>'",
                '  node generate.js --selftest',
                '',
                'input.json keys: ScriptUrl, PageUrl, UserAgent, Bmsz, AcceptLanguage, Abck, IP, Version',
            ].join('\n'));
            process.exit(1);
        }
        let input;
        try { input = JSON.parse(raw); }
        catch (e) { console.error('input is not valid JSON:', e.message); process.exit(1); }
        if (!input.Bmsz) { console.error('input.Bmsz is required (its first numeric token is the cipher seed).'); process.exit(1); }
        process.stdout.write(generate(input));
    }
}

module.exports = { generate };
